src/impl_/coroutine.rs

```diff
@@ -32,17 +32,16 @@ where
 }
 
 fn get_ptr<T: PyClass>(obj: &Py<T>) -> *mut T {
-    // SAFETY: Py<T> can be casted as *const PyCell<T>
-    unsafe { &*(obj.as_ptr() as *const PyCell<T>) }.get_ptr()
+    obj.get_class_object().get_ptr()
 }
 
 pub struct RefGuard<T: PyClass>(Py<T>);
 
 impl<T: PyClass> RefGuard<T> {
     pub fn new(obj: &Bound<'_, PyAny>) -> PyResult<Self> {
-        let owned = obj.downcast::<T>()?;
-        mem::forget(owned.try_borrow()?);
-        Ok(RefGuard(owned.clone().unbind()))
+        let bound = obj.downcast::<T>()?;
+        bound.get_class_object().borrow_checker().try_borrow()?;
```

Comment:
Ah, I see the probem now, `bound.try_borrow` currently clones the `Bound`, so we leak that ref count increment. I guess this could be changed back, once we can store a `&Bound` in `PyRef(mut)`, right?

Reply:
Yes exactly, and potentially we could change it back later. I wonder whether a better follow-up would be to instead change the way this borrow checker works to return a borrow "token" which then needs to be handed back to the borrow checker later (or otherwise panic / abort on drop, maybe just in debug mode). Then what I've written here would be robust and the justification to change back to `forget` a `PyRef` seems weaker.